esl/MongooseIM
# TOML config parser

The goal of this PR is to provide a first POC implemenation of the TOML config parser.

**Done:**
- Simplest code that handles the typical options. All options from the `pgsql_mnesia` preset are supported.
- Both config formats supported at once, TOML takes precendence over the `cfg` file. The paths need to be the same and the configurability is limited on purpose - there is no environment variable for the TOML config format as we mignht end up supporting only one format soon.
- Handle of the happy path when the config file is configured properly.

**To do:**
- Reduce the amount of copy+paste and replacing the ad-hoc structure of the config processing with a more structured approach.
- Refactor the code for improved readability.
- Add or drop the support for config reload (most likely the former).
- Add specs.
- Template the TOML config file for testing and for the release.
- Test the TOML config file on CI.
- Validate the options and handle errors in a unified way.
- If both formats need to be supported, split `mongooseim_config_parser` into two modules: API and the backend which is interchangeable with the TOML parser.
- Use `binary_to_existing_atom` instead of `binary_to_atom`. This is not done yet as some atoms appearing in the config files are not existing yet at the time of parsing.

**Note for the reviewers:**
- Start with `mongooseim.toml` - it's the most important file right now as the format is taking the final shape at the moment.
- If possible, do not mention the things already listed as to do unless there is some important matter.

**How to test it?**
It was tested by uncommenting the `copy` operation for `mongooseim.toml` in `rebar.config`. The tests were performed with:

`./tools/test-runner.sh --one-node --db pgsql redis --skip-small-tests --preset pgsql_mnesia --`

Failing tests - all of them were expected:
- service_mongoose_system_metrics_SUITE - 1 test, requires mim3
- conf_reload_SUITE - 12 tests, reload not supported
- connect_SUITE - 72 tests, reload not supported
- component_SUITE - 21 tests, reload not supported
- login_SUITE - 14 tests, reload not supported
- metrics_api_SUITE - 3 tests, requires mim2
- rest_client_SUITE - 3 tests, requires mim3
- sasl_external_SUITE - 34 tests, reload not supported
